### PR TITLE
OSBuild now uploads container images directly to a container registry

### PR DESCRIPTION
**Pull request by ygalblum**
## CRD
Add a new section in the OSBuildEnvConfig for the container registry parameters
    
## OSBuildEnvCnfig Controller
Worker Config - update the worker configuration file
Job - Pass the secrets (auth and optional CA certificate)
Playbook - Copy the certificates to the VM
Update UT
    
## Config
Update the versions of OSBuild, OSBuild-Worker and OSBuild-Composer
    
## OSBuild Controller
Update composer openapi yaml and client
Use container upload options and status when creating an edge-container
Do not pass nil on optional paramrter
Update UT
    
## Makefile
Add target for updating the composer openapi file
    
## Sample
Update the README along with create_env and the sample

### CODE REVIEW

**Comment by ygalblum** [comment]
This PR is pending the release of the new OSBuild-Worker version v59. Once it's released, the version of OSBuild-Worker, OSBuild and Composer will be updated and the code may be tested E2E

File: README.md [hunk 1]
```diff
@@ -94,6 +94,42 @@ Please note that the provisioning of the OSBuild Operator will also provision th
...
   ```
 
+## Create a Container Registry service
+- Create an HTPasswd file
+  ```bash
+  htpasswd -Bbc /tmp/auth admin <Password>
+  ```
+- Create a secret for the HTPasswd file
+  ```bash
+  kubectl create secret generic container-registry-auth --from-file=/tmp/auth
```
Comment (bardielle) [inline README.md:104]: let's be consistency  kubectl -> oc

File: api/v1alpha1/osbuildenvconfig_types.go [hunk 2]
```diff
@@ -40,6 +40,9 @@ type OSBuildEnvConfigSpec struct {
 	// S3Service holds the configuration needed to connect to the S3 service
 	// +kubebuilder:validation:Required
 	S3Service S3ServiceConfig `json:"s3Service"`
+	// ContainerRegistryService holds the configuration needed to upload container images to the registry
+	// +kubebuilder:validation:Required
+	ContainerRegistryService ContainerRegistryServiceConfig `json:"containerRegistryService"`
```
Comment (bardielle) [inline api/v1alpha1/osbuildenvconfig_types.go:45]: Are you sure it's a mandatory field? what about qcow2 or other image types?

File: api/v1alpha1/osbuildenvconfig_types.go [hunk 3]
```diff
@@ -146,6 +149,27 @@ type GenericS3ServiceConfig struct {
 	SkipSSLVerification *bool `json:"skipSSLVerification,omitempty"`
 }
 
+type ContainerRegistryServiceConfig struct {
```
Comment (bardielle) [inline api/v1alpha1/osbuildenvconfig_types.go:152]: Please add that section to the sample yaml also

Comment (bardielle) [inline api/v1alpha1/osbuildenvconfig_types.go:152]: Please add that field to the sample yaml with setted values from the demo creation

File: internal/conf/config.go [hunk 4]
```diff
@@ -54,10 +54,10 @@ type OperatorConfig struct {
 	EnvoyProxyImageTag string `envconfig:"ENVOY_PROXY_IMAGE_TAG" default:"v1.21-latest"`
 
 	// WorkerOSBuildComposerVersion is the release tag of OSBuild-Composer for the osbuild-worker RPM
-	WorkerOSBuildComposerVersion string `envconfig:"WORKER_OSBUILD_COMPOSER_VERSION" default:"v56"`
+	WorkerOSBuildComposerVersion string `envconfig:"WORKER_OSBUILD_COMPOSER_VERSION" default:"v59"`
```
Comment (bardielle) [inline internal/conf/config.go:57]: Why the version is different from the worker version?

File: README.md (hunk 1 above)
Comment (ygalblum) [reply README.md]: yes, copy paste from my terminal. Fixed

File: resources/templates/worker-config-ansible-playbook.yaml [hunk 5]
```diff
@@ -154,12 +157,40 @@
...
+        register: s3_ca_bundle_file
       - name: Copy the S3 CA Bundle file if exists
         ansible.builtin.copy:
           src: "{{ .OSBuildWorkerS3CABundleDir }}/{{ .OSBuildWorkerS3CABundleKey }}"
           dest: {{"'{{ osbuild_worker_config_directory }}/{{ osbuild_worker_s3_ca_bundle_file }}'"}}
-        when: ca_bundle_file.stat.exists
+        when: s3_ca_bundle_file.stat.exists
+
+  - name: Handle the Container Registry Credential files
+    no_log: True
```
Comment (bardielle) [inline resources/templates/worker-config-ansible-playbook.yaml:168]: Maybe need to set `no_log`to false?

File: resources/templates/worker-setup-job.yaml [hunk 6]
```diff
@@ -35,6 +35,8 @@ spec:
               mountPath: {{ .OSBuildWorkerConfigDir }}
             - name: s3-creds
               mountPath: {{ .OSBuildWorkerS3CredsDir }}
+            - name: cir-creds
```
Comment (bardielle) [inline resources/templates/worker-setup-job.yaml:38]: what is the `i` in the cir for?

File: internal/conf/config.go (hunk 4 above)
Comment (ygalblum) [reply internal/conf/config.go:57]: OSBuild (from which the `osbuild` RPM is release) and OSBuild-Composer (from which the `osbuild-composer-worker` RPM is released) are managed in different repositories, have different release cycles and hence are not on the same version number

File: resources/templates/worker-config-ansible-playbook.yaml (hunk 5 above)
Comment (ygalblum) [reply resources/templates/worker-config-ansible-playbook.yaml:168]: Why? Anything certificate related should not be reflected in the logs (e.g. where they are taken from and to where)

File: resources/templates/worker-setup-job.yaml (hunk 6 above)
Comment (ygalblum) [reply resources/templates/worker-setup-job.yaml:38]: Container Image Registry

File: api/v1alpha1/osbuildenvconfig_types.go (hunk 2 above)
Comment (ygalblum) [reply api/v1alpha1/osbuildenvconfig_types.go:45]: Not providing a Container Registry means the system will not be able to produce edge-container or edge-installer. We can make it optional, but then the OSBuildConfig controller will need to check and fail such requests. Do we want to do that?

Comment (bardielle) [reply api/v1alpha1/osbuildenvconfig_types.go:45]: Mmm not sure what is better
let's keep it like that

**Comment by bardielle** [comment]
What about the changes in the osbuild_controller (postComposer requests) ?

**Comment by ygalblum** [comment]
@bardielle please note that there are additional change to the ones you review. The main part is the changes to the osbuild_controller due to the changes in the API

File: controllers/osbuild_controller_test.go [hunk 7]
```diff
@@ -312,7 +312,7 @@ var _ = Describe("OSBuildEnvConfig Controller", func() {
 
 		It("should requeue if failed to getComposerStatus with error", func() {
 			// given
-			composerClient.EXPECT().GetComposeStatusWithResponse(requestContext, zeroUuid, nil).Return(nil, errFailed)
+			composerClient.EXPECT().GetComposeStatusWithResponse(requestContext, zeroUuid).Return(nil, errFailed)
```
Comment (bardielle) [inline controllers/osbuild_controller_test.go:315]: please add more UT for the new functionality

**Review by bardielle — APPROVED** [review]

File: controllers/osbuild_controller_test.go (hunk 7 above)
Comment (bardielle) [reply controllers/osbuild_controller_test.go:315]: I opened a ticket - https://issues.redhat.com/browse/ECOPROJECT-988

**Pull request merged**